src/hypothesis/internal/conjecture/tree.py

```diff
@@ -0,0 +1,327 @@
+                tree_node[b] = node_index
+                if isinstance(tree_node, DictNode):
+                    tree_node = self.tree[old_node_index] = tree_node.payload
+                if len(tree_node) == 256:
+                    tree_node = self.tree[old_node_index] = array('I', [
+                        tree_node[c] for c in hrange(256)
+                    ])
+            tree_node = self.tree[node_index]
+            if self.__is_dead(node_index):
+                break
+
+        # We don't use this after this point so we might as well get rid of it
+        # to save some memory.
+        del data.capped_indices
+
+        for u, v in data.blocks:
+            # This can happen if we hit a dead node when walking the buffer.
+            # In that case we alrady have this section of the tree mapped.
+            if u >= len(indices):
+                break
+            self.block_sizes[indices[u]] = v - u
+
+        if data.status != Status.OVERRUN and not self.__is_dead(node_index):
+            self.__mark_dead(node_index)
+            if data.status == Status.INTERESTING:
+                self.tree[node_index] = (
+                    data.status, self.origin_cache.setdefault(
+                        data.interesting_origin, data.interesting_origin))
+            else:
+                self.tree[node_index] = data.status
+
+            for j in reversed(indices):
+                if (
+                    len(self.tree[j]) < self.capped.get(j, 255) + 1 and
+                    j not in self.forced
+                ):
+                    break
+                jnode = self.tree[j]
+                if len(jnode) == 256:
+                    values = jnode
+                else:
+                    values = jnode.values()
+                if all(self.__is_dead(v) for v in values):
+                    self.__mark_dead(j)
+                else:
+                    break
+
+    def prescreen_buffer(self, buffer):
+        """Attempt to rule out buffer as a possible interesting candidate.
+
+        Returns False if we know for sure that running this buffer will not
+        produce an interesting result. Returns True if it might (because it
+        explores territory we have not previously tried).
+
+        This is purely an optimisation to try to reduce the number of tests we
+        run. "return True" would be a valid but inefficient implementation.
+
+        """
+        node_index = 0
+        n = len(buffer)
+        for k, b in enumerate(buffer):
+            if self.__is_dead(node_index):
+                return False
+            try:
+                # The block size at that point provides a lower bound on how
+                # many more bytes are required. If the buffer does not have
+                # enough bytes to fulfill that block size then we can rule out
+                # this buffer.
+                if k + self.block_sizes[node_index] > n:
+                    return False
+            except KeyError:
```

Comment:
How "hot" is that code? Because I see failing here with KeyError mostly is ok, but catching an exception is quite expensive in comparison with `__contains__` check - the difference is about the order. May be it worth to switch to LBYL approach to gain a little speedup?

With the same reason you can optimize Nodes `__contains__` to avoid raising exception when you can simply check for a value.